docs/tech/rfc/0019-240310-cesium-race-conditions.md

```diff
@@ -7,129 +7,616 @@
+in the file controller under a `RLock` and upon finding a file within the size limit, we
+try to acquire it by comparing-and-swapping its flag (`True` indicates the writer is currently
+in use) from `False` to `True`. If successfully acquired, we return that file handle as
+our writer and the acquisition of a writer is complete.
+Note that although `RLock` does not guarantee exclusivity, the atomic flag
+guarantees that only one writer will ever be in control of a file handle.
+
+If none were acquired out of the open file handles, we first calculate the number of file
+handles in the file controller under an `RLock`: if it is below the limit, then we may
+acquire a new writer – note that there is a possible race condition here (#1): multiple
+callers may each try to create a new writer after checking that the number of file handles
+is below the limit, pushing the total number of file handles beyond the limit.
+
+In the case where we open a new writer, we lock the mutex of the writer completely with
+an exclusive `Lock` and release it when we have a new writer. The acquisition of a writer
+is complete.
+
+In the case where we do not open a new writer, i.e. the number of file descriptors exceeds
+the limit, we garbage-collect all writers (file handles to oversize files) under an exclusive
+lock, and try to acquire a writer again if we garbage-collected any writers. If we did
+not, then we wait for a writer to be released (signaled by an input into the channel `release`),
+and try to acquire a writer again.
+
+#### 4.2.1.2 Writing data with a writer
+
+`Write` writes binary data into the file handle acquired in the previous step and updates
+the `len` and `fileSize` fields of the writer. Note that `Write` may not be called
+concurrently with any other Writer methods because it modifies these fields and writes to
+the file system.
+
+#### 4.2.1.3 Committing the data in a writer
+
+`Commit` first reads the length of the internal file handle to determine how much data
+was written in this domain: note that there is a race condition here: the internal
+file handle's length may get changed via another `Commit`, which may assign `internal` a
+new writer should there be a file switch, causing `w.internal.len()` to be 0 instead of
+the actual length.
+
+Using the `len` field, the writer determines whether a file cutoff is needed. Note that
+this step is race-free because `len` may be only updated via `Write`, which is not to be
+called concurrently.
+
+Using the length and offset information of the `TrackedWriterCloser`, the index is
+inserted with the new pointer under an exclusive lock, therefore race-free. After the
+insertion, the writer may switch files, changing the `fileKey`, `internal`, `fileSize`
+fields.
+
+Note that these race conditions resulting from concurrent calls to `Commit` should never
+happen, as the method is documented to not be called concurrently.
```

Comment:
What does this mean? The comment above the commit method just says "don't call this function concurrently"?

Reply:
yes